File: python/modelcreator/java/model/events/EventFactory.java
package model.events;

import java.io.FileNotFoundException;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;

import javax.xml.stream.XMLStreamException;

import model.SBMLUtils;

import org.sbml.jsbml.ASTNode;
import org.sbml.jsbml.Event;
import org.sbml.jsbml.Model;
import org.sbml.jsbml.SBMLDocument;
import org.sbml.jsbml.SBMLException;
import org.sbml.jsbml.SBMLWriter;
import org.sbml.jsbml.Trigger;
import org.sbml.jsbml.text.parser.ParseException;

/**
 * Handeling of model events for the model.
 *
 * @author Matthias Koenig
 * @date 2014-04-19
 */

public class EventFactory {
	
	/** Creates the events based on the given EventData. 
	 * @throws ParseException */
	public static void createEventsInModel(Collection<EventData> eDataCol, Model model) throws ParseException{
		for (EventData eData: eDataCol){
			createEventInModelFromEventData(eData, model);
		}
	}
	 
	private static Event createEventInModelFromEventData(EventData eData, Model model) throws ParseException{
		Event e = model.createEvent(eData.getId());
		
		e.setName(eData.getName());
		e.setUseValuesFromTriggerTime(true);
		Trigger t = e.createTrigger(true, true);
		t.setMath(ASTNode.parseFormula(eData.getTrigger()));
		HashMap<String, String> assignments = eData.getAssignments();
		for (String key : assignments.keySet()){
			String a = assignments.get(key);
			e.createEventAssignment(key, ASTNode.parseFormula(a));	
		}
		return e;
	}
	
	/** Creates a dilution peak in the periportal components.
	 * The peak area is normalized to 1, i.e height = 1/duration.
	 * Returns the EventData for the dilution peak.
	 * TODO: some testing and quality checking of components
	 */
	public static Collection<EventData> createDilutionPeakEventData(Double tstart, Double duration){
		Double tend = (tstart + duration);
		Double p = 1.0/duration;
		String[] compounds = {"PP__gal", "PP__rbcM", "PP__alb", "PP__h2oM", "PP__suc"};
		Double[] peak = {p, p, p, p, p};
		Double[] background = {0.0, 0.0, 0.0, 0.0, 0.0};
		
		return createPeakEventData(tstart, tend, compounds, peak, background);
	}
	
	public static Collection<EventData> createPeakEventData(Double tstart, Double tend, String[] compounds, 
					Double[] peak, Double[] background){
		Collection<EventData> eDataCol = new LinkedList<EventData>();

		// Create peak events
		HashMap<String, String> aPre = new HashMap<String, String>();
		HashMap<String, String> aPeak = new HashMap<String, String>();
		HashMap<String, String> aPost = new HashMap<String, String>();
		for (int k=0; k<compounds.length; ++k){
			aPre.put(compounds[k], background[k].toString());
			aPeak.put(compounds[k], peak[k].toString());
			aPost.put(compounds[k], background[k].toString());
		}
		EventData eD0 = new EventData("EDIL_0", "pre Dilution Peak [PP]",
				   createTimeString(0.0), aPre);
		eDataCol.add(eD0);
		EventData eD1 = new EventData("EDIL_1", "Dilution Peak [PP]",
									   createTimeString(tstart), aPeak);
		eDataCol.add(eD1);
		EventData eD2 = new EventData("EDIL_2", "post Dilution Peak [PP]",
						createTimeString(tend), aPost);
		eDataCol.add(eD2);
		return eDataCol;
	}
	
	private static String createTimeString(Double time){
		return String.format("(time >= %f)", time);
	}
	
	
	public static void main(String[] args) throws ParseException, SBMLException, XMLStreamException, FileNotFoundException{
		
		Model model = new Model(SBMLUtils.SBML_LEVEL, SBMLUtils.SBML_VERSION);	
		model.createSpecies("PP__gal");
		model.createSpecies("PP__rbcM");
		model.createSpecies("PP__alb");
		model.createSpecies("PP__h2oM");
		model.createSpecies("PP__suc");
		
		Collection<EventData> eDataCol = createDilutionPeakEventData(10.0, 0.5);
		createEventsInModel(eDataCol, model);
		
		SBMLDocument doc = new SBMLDocument(SBMLUtils.SBML_LEVEL, SBMLUtils.SBML_VERSION);
		doc.setModel(model);
		(new SBMLWriter()).write(doc, System.out);
		(new SBMLWriter()).write(doc, "testEvents.xml");
	}
	
}

File: python/modelcreator/creator/events/EventFactory.py
'''
Creating the simulation event data.
The EventData is used to generate the SBML events for the model.

Created on Jul 2, 2014
@author: mkoenig
'''

class EventData():
    def __init__(self, eid, name, trigger, assignments):
        self.eid = eid
        self.name = name
        self.trigger = trigger
        self.assignments = assignments
    
    def info(self):
        print '-' * 20
        print self.eid
        print self.name
        print self.trigger
        print self.assignments
        
        
        
def createDilutionEventData(time_start, duration):
    species = ["PP__galM", "PP__rbcM", "PP__alb", "PP__h2oM", "PP__suc"]
    # all species have the same peak height based on the duration
    base = ('{} mM'.format(0.0), ) * len(species)
    peak = ('{} mM'.format(1.0/duration),) * len(species);
    return createPeakEventData(species, base, peak, time_start=time_start, duration=duration)
        
        
def createPeakEventData(species, base, peak, time_start, duration):
    ''' 
    Creates a dilution peak in the given species beginning at the
    start time and with the provided duration.
    '''
    time_end = time_start + duration
    ed1 = EventData("EDIL_0", "pre peak [PP]",
                   createTriggerFromTime(0.0), createAssignmentsDict(species, base))
    ed2 = EventData("EDIL_1", "peak [PP]",
                   createTriggerFromTime(time_start), createAssignmentsDict(species, peak))
    ed3 = EventData("EDIL_2", "post peak [PP]",
                   createTriggerFromTime(time_end), createAssignmentsDict(species, base))

    return [ed1, ed2, ed3]

def createGalactoseChallengeEventData(tc_start = 100.0, peak_var='gal_challenge'):
    ed = EventData("EDIL_0", "galactose challenge",
                   createTriggerFromTime(tc_start), {'PP__gal': peak_var})
    return [ed]

def createGalactoseStepEventData():
    ''' Stepwise increase in PP__gal over time.'''
    event_data = []
    duration = 1000.0;
    for k in range(0,21):
        time = 0.0 + k*duration;
        gal = 0.0 + k*0.5
        ed = EventData('ESTEP_{}'.format(k), "galactose step",
                   createTriggerFromTime(time), {'PP__gal': '{} mM'.format(gal)})
        event_data.append(ed)
    return event_data


def createTriggerFromTime(t):
    return '(time >= {})'.format(t)

def createAssignmentsDict(species, values):
    return dict(zip(species, values) )

#####################################################################
if __name__ == '__main__':
    elist = createDilutionEventData()
    for edata in elist:
        edata.info()
    
    elist = createDilutionEventData(tp_start=100.0, duration=10.0)
    for edata in elist:
        edata.info()
    
    elist = createGalactoseStepEventData()
    for edata in elist:
        edata.info()
    
    
    
    